File: backend/schemas/analysis.py
from __future__ import annotations

from pydantic import BaseModel


class SummaryStats(BaseModel):
    mean: float | None = None
    std: float | None = None
    min: float | None = None
    max: float | None = None
    q1: float | None = None
    median: float | None = None
    q3: float | None = None
    skewness: float | None = None


class CorrelationMatrix(BaseModel):
    columns: list[str]
    values: list[list[float | None]]


class OutlierResult(BaseModel):
    column: str
    lower_bound: float
    upper_bound: float
    outlier_count: int
    outlier_indices: list[int]


class QualityAlert(BaseModel):
    column: str
    alert_type: str  # "constant" | "high_cardinality" | "high_null" | "high_skew"
    message: str
    severity: str  # "warning" | "info"


class AnalysisResultResponse(BaseModel):
    summary: dict[str, SummaryStats]
    correlation: CorrelationMatrix
    outliers: list[OutlierResult]
    quality_alerts: list[QualityAlert]
    row_count: int
    column_count: int
    missing_ratio: float
    duplicate_count: int

File: frontend/src/types/analysis.ts
/**
 * Analysis result types — mirrors backend/schemas/analysis.py (Pydantic v2).
 * snake_case preserved (no camelCase transform).
 * Must remain in sync with backend response schemas.
 */

export interface SummaryStats {
  mean: number | null;
  std: number | null;
  min: number | null;
  max: number | null;
  q1: number | null;
  median: number | null;
  q3: number | null;
  skewness: number | null;
}

export interface CorrelationMatrix {
  columns: string[];
  values: (number | null)[][];
}

export interface OutlierResult {
  column: string;
  lower_bound: number;
  upper_bound: number;
  outlier_count: number;
  outlier_indices: number[];
}

export interface QualityAlert {
  column: string;
  alert_type: string;
  message: string;
  severity: string;
}

export interface AnalysisResultResponse {
  summary: Record<string, SummaryStats>;
  correlation: CorrelationMatrix;
  outliers: OutlierResult[];
  quality_alerts: QualityAlert[];
  row_count: number;
  column_count: number;
  missing_ratio: number;
  duplicate_count: number;
}
